File: students/Rro0O/homework-4/src/bankingSystem.js
/**
 * SecureBank online banking system: JavaScript port of src/banking_system.py.
 *
 * Every operation returns a result object instead of throwing:
 *   { success: boolean, error: string | null, errors: string[], ... }
 * Money is computed in integer cents to avoid floating point drift.
 */

const ACCOUNT_TYPES = {
  Savings: { minBalance: 100, monthlyFee: 5, feeWaiverAbove: 1000, dailyLimit: 2000 },
  Checking: { minBalance: 0, monthlyFee: 10, feeWaiverAbove: 5000, dailyLimit: 5000 },
  Premium: { minBalance: 10000, monthlyFee: 0, feeWaiverAbove: 0, dailyLimit: 50000 },
};

const STATE_ACTIVE = 'Active';
const STATE_FROZEN = 'Frozen';
const STATE_SUSPENDED = 'Suspended';
const STATE_CLOSED = 'Closed';

// States in which money may move. Suspended accounts keep working (with a warning).
const OPERABLE_STATES = [STATE_ACTIVE, STATE_SUSPENDED];

const VALID_PAYEES = {
  'CFE Electricity': 'utilities',
  'Telmex Internet': 'utilities',
  'SIAPA Water': 'utilities',
  'Visa Credit Card': 'credit card',
  'Mastercard Credit Card': 'credit card',
};

const CSV_HEADER = ['date', 'type', 'amount', 'balance_after', 'description'];

const pad = (number) => String(number).padStart(2, '0');

/** Convert dollars to integer cents. */
const toCents = (dollars) => Math.round(dollars * 100);

/** Convert integer cents to dollars. */
const fromCents = (cents) => cents / 100;

/**
 * Normalise a Date or 'YYYY-MM-DD' string to an ISO day string; null if invalid.
 * @param {Date|string} value
 * @returns {string|null}
 */
function parseDate(value) {
  if (value instanceof Date && !Number.isNaN(value.getTime())) {
    return `${value.getFullYear()}-${pad(value.getMonth() + 1)}-${pad(value.getDate())}`;
  }
  if (typeof value === 'string' && /^\d{4}-\d{2}-\d{2}$/.test(value)) {
    const [year, month, day] = value.split('-').map(Number);
    const check = new Date(Date.UTC(year, month - 1, day));
    const isRealDate =
      check.getUTCFullYear() === year && check.getUTCMonth() === month - 1 && check.getUTCDate() === day;
    return isRealDate ? value : null;
  }
  return null;
}

/** @returns {string} today's date as an ISO day string */
function today() {
  return parseDate(new Date());
}

/**
 * Validate a monetary amount.
 * @param {*} amount
 * @returns {{cents: number|null, error: string|null}}
 */
function validateAmount(amount) {
  if (typeof amount !== 'number' || !Number.isFinite(amount)) {
    return { cents: null, error: 'Amount must be a number' };
  }
  if (amount <= 0) {
    return { cents: null, error: 'Amount must be positive' };
  }
  const scaled = amount * 100;
  if (Math.abs(scaled - Math.round(scaled)) > 1e-6) {
    return { cents: null, error: 'Amount must have at most 2 decimal places' };
  }
  return { cents: Math.round(scaled), error: null };
}

const fail = (...errors) => ({ success: false, error: errors[0], errors });
const ok = (extra = {}) => ({ success: true, error: null, errors: [], ...extra });

const csvField = (value) => {
  const text = String(value);
  return /[",\n]/.test(text) ? `"${text.replace(/"/g, '""')}"` : text;
};

class BankAccount {
  /**
   * @param {string} accountType Savings | Checking | Premium
   * @param {number} initialBalance
   * @param {string} [owner]
   * @param {string|null} [email]
   */
  constructor(accountType, initialBalance, owner = 'Customer', email = null) {
    if (!Object.prototype.hasOwnProperty.call(ACCOUNT_TYPES, accountType)) {
      throw new Error(`Invalid account type: '${accountType}'`);
    }
    if (typeof initialBalance !== 'number' || Number.isNaN(initialBalance)) {
      throw new Error('Initial balance must be a number');
    }
    const minimum = ACCOUNT_TYPES[accountType].minBalance;
    if (initialBalance < 0) {
      throw new Error('Initial balance cannot be negative');
    }
    if (initialBalance < minimum) {
      throw new Error(`Initial balance is below the ${accountType} minimum of ${minimum}`);
    }
    if (typeof owner !== 'string' || owner.trim() === '') {
      throw new Error('Owner name is required');
    }
    this.accountType = accountType;
    this.balance = initialBalance;
    this.owner = owner.trim();
    this.email = email;
    this.state = STATE_ACTIVE;
    this.dailyTransferTotal = 0;
    this.unpaidFees = 0;
    this.scheduledPayments = [];
    this.transactions = [];
    this.lastActivityDay = null;
  }

  // ------------------------------------------------------------------ helpers
  getDailyLimit() {
    return ACCOUNT_TYPES[this.accountType].dailyLimit;
  }

  getMinimumBalance() {
    return ACCOUNT_TYPES[this.accountType].minBalance;
  }

  resetDailyTotalIfNewDay(day) {
    if (this.lastActivityDay !== day) {
      this.dailyTransferTotal = 0;
      this.lastActivityDay = day;
    }
  }

  record(day, type, amount, description) {
    this.transactions.push({ date: day, type, amount, balanceAfter: this.balance, description });
  }

  /** Apply Active <-> Suspended transitions; returns a warning when the account was just suspended. */
  refreshStateAfterBalanceChange() {
    const belowMinimum = this.balance < this.getMinimumBalance();
    if (this.state === STATE_ACTIVE && belowMinimum) {
      this.state = STATE_SUSPENDED;
      return 'Balance is below the minimum: account suspended';
    }
    if (this.state === STATE_SUSPENDED && !belowMinimum) {
      this.state = STATE_ACTIVE;
    }
    return null;
  }

  stateError() {
    if (this.state === STATE_FROZEN) return 'Account is frozen';
    if (this.state === STATE_CLOSED) return 'Account is closed';
    return null;
  }

  debit(cents, day, type, description) {
    this.balance = fromCents(toCents(this.balance) - cents);
    const warning = this.refreshStateAfterBalanceChange();
    this.record(day, type, -fromCents(cents), description);
    return warning;
  }

  // ------------------------------------------------------------------- money
  /**
   * Transfer money out of this account. Errors are collected so the decision table can be checked rule by rule.
   * @param {number} amount
   * @param {{destination?: BankAccount|string|null, onDate?: Date|string}} [options]
   */
  transfer(amount, { destination = null, onDate } = {}) {
    const day = parseDate(onDate) || today();
    const stateError = this.stateError();
    if (stateError) return fail(stateError);
    const { cents, error } = validateAmount(amount);
    if (error) return fail(error);
    if (destination === this) return fail('Cannot transfer to the same account');

    this.resetDailyTotalIfNewDay(day);
    const errors = [];
    if (toCents(this.dailyTransferTotal) + cents > toCents(this.getDailyLimit())) {
      errors.push('Exceeds daily limit');
    }
    if (cents > toCents(this.balance)) {
      errors.push('Insufficient funds');
    }
    const isAccount = destination instanceof BankAccount;
    if (isAccount && !OPERABLE_STATES.includes(destination.state)) {
      errors.push('Destination account cannot receive funds');
    }
    if (errors.length > 0) return fail(...errors);

    this.dailyTransferTotal = fromCents(toCents(this.dailyTransferTotal) + cents);
    const label = isAccount ? destination.owner : destination || 'external account';
    const warning = this.debit(cents, day, 'transfer', `Transfer to ${label}`);
    if (isAccount) {
      destination.deposit(fromCents(cents), { onDate: day, description: `Transfer from ${this.owner}` });
    }
    return ok({ warning, state: this.state, balance: this.balance });
  }

  /** Add money; a deposit that restores the minimum reactivates a Suspended account. */
  deposit(amount, { onDate, description = 'Deposit' } = {}) {
    const day = parseDate(onDate) || today();
    const stateError = this.stateError();
    if (stateError) return fail(stateError);
    const { cents, error } = validateAmount(amount);
    if (error) return fail(error);
    this.balance = fromCents(toCents(this.balance) + cents);
    this.refreshStateAfterBalanceChange();
    this.record(day, 'deposit', fromCents(cents), description);
    return ok({ state: this.state, balance: this.balance });
  }

  /** Pay a registered payee now, or schedule the payment for a future date. */
  payBill(payee, amount, { paymentDate, onDate } = {}) {
    const current = parseDate(onDate) || today();
    const stateError = this.stateError();
    if (stateError) return fail(stateError);
    const errors = [];
    if (typeof payee !== 'string' || payee.trim() === '') {
      errors.push('Payee is required');
    } else if (!Object.prototype.hasOwnProperty.call(VALID_PAYEES, payee.trim())) {
      errors.push('Invalid payee');
    }
    const { cents, error } = validateAmount(amount);
    if (error) {
      errors.push(error);
    } else if (cents > toCents(this.balance)) {
      errors.push('Insufficient funds');
    }
    const due = paymentDate === undefined || paymentDate === null ? current : parseDate(paymentDate);
    if (due === null) {
      errors.push('Invalid payment date');
    } else if (due < current) {
      errors.push('Payment date cannot be in the past');
    }
    if (errors.length > 0) return fail(...errors);

    const name = payee.trim();
    if (due > current) {
      this.scheduledPayments.push({ payee: name, amount: fromCents(cents), date: due });
      return ok({ scheduled: true, paymentDate: due, balance: this.balance });
    }
    const warning = this.debit(cents, current, 'bill payment', `Bill payment to ${name}`);
    return ok({ scheduled: false, warning, state: this.state, balance: this.balance });
  }

  /** Charge the monthly fee (only on the 1st of the month, waived above the threshold). */
  processMonthlyFee(onDate) {
    const day = parseDate(onDate);
    if (day === null) return fail('Invalid date');
    if (!OPERABLE_STATES.includes(this.state)) return fail(`Fees are not processed for ${this.state} accounts`);
    if (!day.endsWith('-01')) return fail('Monthly fees are only charged on the 1st of the month');

    const rules = ACCOUNT_TYPES[this.accountType];
    const feeCents = toCents(rules.monthlyFee);
    if (feeCents === 0 || this.balance > rules.feeWaiverAbove) {
      return ok({ feeCharged: 0, waived: true, state: this.state, balance: this.balance });
    }
    if (toCents(this.balance) < feeCents) {
      this.unpaidFees = fromCents(toCents(this.unpaidFees) + feeCents);
      this.state = STATE_SUSPENDED;
      return ok({ feeCharged: 0, waived: false, suspended: true, state: this.state, balance: this.balance });
    }
    const warning = this.debit(feeCents, day, 'fee', 'Monthly fee');
    return ok({ feeCharged: fromCents(feeCents), waived: false, warning, state: this.state, balance: this.balance });
  }

  // --------------------------------------------------------- account lifecycle
  /** Active -> Frozen (customer request or fraud detection). */
  freeze(reason = 'customer request') {
    if (this.state !== STATE_ACTIVE) return fail(`Cannot freeze an account in state ${this.state}`);
    this.state = STATE_FROZEN;
    return ok({ state: this.state, reason });
  }

  /** Frozen -> Active. */
  unfreeze() {
    if (this.state !== STATE_FROZEN) return fail(`Cannot unfreeze an account in state ${this.state}`);
    this.state = STATE_ACTIVE;
    return ok({ state: this.state });
  }

  /** Any state -> Closed. Closed accounts can never be reopened. */
  close() {
    if (this.state === STATE_CLOSED) return fail('Account is closed');
    this.state = STATE_CLOSED;
    const finalStatement = { owner: this.owner, finalBalance: this.balance, transactions: this.transactions.length };
    return ok({ state: this.state, finalStatement });
  }

  /** Closed accounts cannot be reopened; every other state is left untouched. */
  reopen() {
    return fail(this.state === STATE_CLOSED ? 'Account is closed' : 'Account is not closed');
  }

  /** Update the owner name and/or e-mail. Not allowed on closed accounts. */
  updateInfo({ owner, email } = {}) {
    if (this.state === STATE_CLOSED) return fail('Account is closed');
    const errors = [];
    if (owner !== undefined && (typeof owner !== 'string' || owner.trim() === '')) {
      errors.push('Owner name is required');
    }
    if (email !== undefined && (typeof email !== 'string' || !email.includes('@'))) {
      errors.push('Invalid e-mail address');
    }
    if (errors.length > 0) return fail(...errors);
    if (owner !== undefined) this.owner = owner.trim();
    if (email !== undefined) this.email = email;
    return ok({ owner: this.owner, email: this.email });
  }

  /** View the balance (allowed in every state, including Frozen). */
  getBalance() {
    return this.balance;
  }

  // ------------------------------------------------------------ history / CSV
  /**
   * Transactions between start and end (both inclusive, both optional).
   * Throws for unparsable dates or when start is after end.
   */
  getHistory(start, end) {
    const first = start === undefined || start === null ? null : parseDate(start);
    const last = end === undefined || end === null ? null : parseDate(end);
    const startGiven = start !== undefined && start !== null;
    const endGiven = end !== undefined && end !== null;
    if ((startGiven && first === null) || (endGiven && last === null)) {
      throw new Error('Invalid date');
    }
    if (first && last && first > last) {
      throw new Error('Start date must not be after end date');
    }
    return this.transactions.filter((tx) => (!first || tx.date >= first) && (!last || tx.date <= last));
  }

  /** Export the (optionally filtered) history as CSV text. */
  exportCsv(start, end) {
    const rows = this.getHistory(start, end).map((tx) =>
      [tx.date, tx.type, tx.amount.toFixed(2), tx.balanceAfter.toFixed(2), tx.description].map(csvField).join(','),
    );
    return `${[CSV_HEADER.join(','), ...rows].join('\n')}\n`;
  }
}

module.exports = {
  BankAccount,
  ACCOUNT_TYPES,
  VALID_PAYEES,
  parseDate,
  validateAmount,
};

File: students/Rro0O/homework-4/src/banking_system.py
"""SecureBank online banking system: the system under test for homework 4.

Business rules implemented here come straight from the assignment statement:
account types (Savings / Checking / Premium), account states (Active / Frozen /
Suspended / Closed), transfers, bill payments, monthly fees and a filterable
transaction history that can be exported to CSV.

Every operation returns a result dictionary instead of raising, so callers (and
the black box tests) only depend on the observable behaviour:

    {"success": bool, "error": str | None, "errors": [str, ...], ...}
"""

import csv
import io
from datetime import date, datetime
from decimal import Decimal

ACCOUNT_TYPES = {
    "Savings": {"min_balance": 100, "monthly_fee": 5, "fee_waiver_above": 1000, "daily_limit": 2000},
    "Checking": {"min_balance": 0, "monthly_fee": 10, "fee_waiver_above": 5000, "daily_limit": 5000},
    "Premium": {"min_balance": 10000, "monthly_fee": 0, "fee_waiver_above": 0, "daily_limit": 50000},
}

STATE_ACTIVE = "Active"
STATE_FROZEN = "Frozen"
STATE_SUSPENDED = "Suspended"
STATE_CLOSED = "Closed"

# States in which money may move. Suspended accounts keep working (with a warning).
OPERABLE_STATES = (STATE_ACTIVE, STATE_SUSPENDED)

MIN_TRANSFER = Decimal("0.01")

VALID_PAYEES = {
    "CFE Electricity": "utilities",
    "Telmex Internet": "utilities",
    "SIAPA Water": "utilities",
    "Visa Credit Card": "credit card",
    "Mastercard Credit Card": "credit card",
}

CSV_HEADER = ["date", "type", "amount", "balance_after", "description"]


def parse_date(value):
    """Return a ``date`` for a date, datetime or ISO ``YYYY-MM-DD`` string; ``None`` if invalid."""
    if isinstance(value, datetime):
        return value.date()
    if isinstance(value, date):
        return value
    if isinstance(value, str):
        try:
            return datetime.strptime(value, "%Y-%m-%d").date()
        except ValueError:
            return None
    return None


def validate_amount(amount):
    """Validate a monetary amount.

    Returns ``(Decimal | None, error | None)``. Booleans, strings, NaN/infinity and
    values with more than two decimal places are rejected.
    """
    if isinstance(amount, bool) or not isinstance(amount, (int, float, Decimal)):
        return None, "Amount must be a number"
    value = Decimal(str(amount))
    if not value.is_finite():
        return None, "Amount must be a number"
    if value <= 0:
        return None, "Amount must be positive"
    if value != value.quantize(MIN_TRANSFER):
        return None, "Amount must have at most 2 decimal places"
    return value, None


def _money(value):
    """Convert a Decimal to a float rounded to cents."""
    return float(round(value, 2))


def _fail(*errors, **extra):
    """Build a failed result; ``error`` is the first (highest priority) message."""
    result = {"success": False, "error": errors[0], "errors": list(errors)}
    result.update(extra)
    return result


def _ok(**extra):
    """Build a successful result."""
    result = {"success": True, "error": None, "errors": []}
    result.update(extra)
    return result


class BankAccount:  # pylint: disable=too-many-instance-attributes
    """A SecureBank account with its balance, state and transaction history."""

    def __init__(self, account_type, initial_balance, owner="Customer", email=None):
        if account_type not in ACCOUNT_TYPES:
            raise ValueError(f"Invalid account type: {account_type!r}")
        if isinstance(initial_balance, bool) or not isinstance(initial_balance, (int, float, Decimal)):
            raise ValueError("Initial balance must be a number")
        minimum = ACCOUNT_TYPES[account_type]["min_balance"]
        if initial_balance < 0:
            raise ValueError("Initial balance cannot be negative")
        if initial_balance < minimum:
            raise ValueError(f"Initial balance is below the {account_type} minimum of {minimum}")
        if not isinstance(owner, str) or not owner.strip():
            raise ValueError("Owner name is required")
        self.account_type = account_type
        self.balance = float(initial_balance)
        self.owner = owner.strip()
        self.email = email
        self.state = STATE_ACTIVE
        self.daily_transfer_total = 0.0
        self.unpaid_fees = 0.0
        self.scheduled_payments = []
        self.transactions = []
        self._last_activity_day = None

    # ------------------------------------------------------------------ helpers
    def get_daily_limit(self):
        """Return the daily transfer limit for this account type."""
        return ACCOUNT_TYPES.get(self.account_type, {}).get("daily_limit", 0)

    def get_minimum_balance(self):
        """Return the minimum balance for this account type."""
        return ACCOUNT_TYPES[self.account_type]["min_balance"]

    def _reset_daily_total_if_new_day(self, day):
        """The daily limit resets at midnight, i.e. whenever the calendar day changes."""
        if self._last_activity_day != day:
            self.daily_transfer_total = 0.0
            self._last_activity_day = day

    def _record(self, day, kind, amount, description):
        self.transactions.append(
            {
                "date": day,
                "type": kind,
                "amount": amount,
                "balance_after": self.balance,
                "description": description,
            }
        )

    def _refresh_state_after_balance_change(self):
        """Apply the Active <-> Suspended transitions caused by balance changes.

        Returns a warning message when the account has just become Suspended.
        """
        below_minimum = self.balance < self.get_minimum_balance()
        if self.state == STATE_ACTIVE and below_minimum:
            self.state = STATE_SUSPENDED
            return "Balance is below the minimum: account suspended"
        if self.state == STATE_SUSPENDED and not below_minimum:
            self.state = STATE_ACTIVE
        return None

    def _state_error(self):
        """Return the error for accounts that cannot move money, else ``None``."""
        if self.state == STATE_FROZEN:
            return "Account is frozen"
        if self.state == STATE_CLOSED:
            return "Account is closed"
        return None

    def _debit(self, amount, day, kind, description):
        """Remove ``amount`` (Decimal) from the balance and record it. Returns a warning or ``None``."""
        self.balance = _money(Decimal(str(self.balance)) - amount)
        warning = self._refresh_state_after_balance_change()
        self._record(day, kind, -_money(amount), description)
        return warning

    # ------------------------------------------------------------------- money
    def transfer(self, amount, destination=None, on_date=None):
        """Transfer money out of this account.

        ``destination`` may be another ``BankAccount`` (own accounts), an external
        account number (another bank) or ``None``. Errors are collected so the
        decision table can be checked rule by rule.
        """
        day = parse_date(on_date) or date.today()
        state_error = self._state_error()
        if state_error:
            return _fail(state_error)
        value, amount_error = validate_amount(amount)
        if amount_error:
            return _fail(amount_error)
        if destination is self:
            return _fail("Cannot transfer to the same account")

        self._reset_daily_total_if_new_day(day)
        errors = []
        if Decimal(str(self.daily_transfer_total)) + value > self.get_daily_limit():
            errors.append("Exceeds daily limit")
        if value > Decimal(str(self.balance)):
            errors.append("Insufficient funds")
        if isinstance(destination, BankAccount) and destination.state not in OPERABLE_STATES:
            errors.append("Destination account cannot receive funds")
        if errors:
            return _fail(*errors)

        self.daily_transfer_total = _money(Decimal(str(self.daily_transfer_total)) + value)
        label = destination.owner if isinstance(destination, BankAccount) else (destination or "external account")
        warning = self._debit(value, day, "transfer", f"Transfer to {label}")
        if isinstance(destination, BankAccount):
            destination.deposit(float(value), on_date=day, description=f"Transfer from {self.owner}")
        return _ok(warning=warning, state=self.state, balance=self.balance)

    def deposit(self, amount, on_date=None, description="Deposit"):
        """Add money to the account. A deposit that restores the minimum reactivates a Suspended account."""
        day = parse_date(on_date) or date.today()
        state_error = self._state_error()
        if state_error:
            return _fail(state_error)
        value, amount_error = validate_amount(amount)
        if amount_error:
            return _fail(amount_error)
        self.balance = _money(Decimal(str(self.balance)) + value)
        self._refresh_state_after_balance_change()
        self._record(day, "deposit", _money(value), description)
        return _ok(state=self.state, balance=self.balance)

    def pay_bill(self, payee, amount, payment_date=None, on_date=None):
        """Pay a registered payee now, or schedule the payment for a future date."""
        today = parse_date(on_date) or date.today()
        state_error = self._state_error()
        if state_error:
            return _fail(state_error)
        errors = []
        if not isinstance(payee, str) or not payee.strip():
            errors.append("Payee is required")
        elif payee.strip() not in VALID_PAYEES:
            errors.append("Invalid payee")
        value, amount_error = validate_amount(amount)
        if amount_error:
            errors.append(amount_error)
        elif value > Decimal(str(self.balance)):
            errors.append("Insufficient funds")
        due = today if payment_date is None else parse_date(payment_date)
        if due is None:
            errors.append("Invalid payment date")
        elif due < today:
            errors.append("Payment date cannot be in the past")
        if errors:
            return _fail(*errors)

        payee = payee.strip()
        if due > today:
            self.scheduled_payments.append({"payee": payee, "amount": _money(value), "date": due})
            return _ok(scheduled=True, payment_date=due, balance=self.balance)
        warning = self._debit(value, today, "bill payment", f"Bill payment to {payee}")
        return _ok(scheduled=False, warning=warning, state=self.state, balance=self.balance)

    def process_monthly_fee(self, on_date):
        """Charge the monthly fee (only on the 1st of the month, waived above the threshold)."""
        day = parse_date(on_date)
        if day is None:
            return _fail("Invalid date")
        if self.state not in OPERABLE_STATES:
            return _fail(f"Fees are not processed for {self.state} accounts")
        if day.day != 1:
            return _fail("Monthly fees are only charged on the 1st of the month")

        rules = ACCOUNT_TYPES[self.account_type]
        fee = Decimal(str(rules["monthly_fee"]))
        waived = fee == 0 or self.balance > rules["fee_waiver_above"]
        if waived:
            return _ok(fee_charged=0.0, waived=True, state=self.state, balance=self.balance)
        if Decimal(str(self.balance)) < fee:
            self.unpaid_fees = _money(Decimal(str(self.unpaid_fees)) + fee)
            self.state = STATE_SUSPENDED
            return _ok(fee_charged=0.0, waived=False, suspended=True, state=self.state, balance=self.balance)
        warning = self._debit(fee, day, "fee", "Monthly fee")
        return _ok(fee_charged=float(fee), waived=False, warning=warning, state=self.state, balance=self.balance)

    # --------------------------------------------------------- account lifecycle
    def freeze(self, reason="customer request"):
        """Active -> Frozen (customer request or fraud detection)."""
        if self.state != STATE_ACTIVE:
            return _fail(f"Cannot freeze an account in state {self.state}")
        self.state = STATE_FROZEN
        return _ok(state=self.state, reason=reason)

    def unfreeze(self):
        """Frozen -> Active."""
        if self.state != STATE_FROZEN:
            return _fail(f"Cannot unfreeze an account in state {self.state}")
        self.state = STATE_ACTIVE
        return _ok(state=self.state)

    def close(self):
        """Any state -> Closed. Closed accounts can never be reopened."""
        if self.state == STATE_CLOSED:
            return _fail("Account is closed")
        self.state = STATE_CLOSED
        statement = {"owner": self.owner, "final_balance": self.balance, "transactions": len(self.transactions)}
        return _ok(state=self.state, final_statement=statement)

    def reopen(self):
        """Closed accounts cannot be reopened; every other state is left untouched."""
        if self.state == STATE_CLOSED:
            return _fail("Account is closed")
        return _fail("Account is not closed")

    def update_info(self, owner=None, email=None):
        """Update the owner name and/or e-mail. Not allowed on closed accounts."""
        if self.state == STATE_CLOSED:
            return _fail("Account is closed")
        errors = []
        if owner is not None and (not isinstance(owner, str) or not owner.strip()):
            errors.append("Owner name is required")
        if email is not None and (not isinstance(email, str) or "@" not in email):
            errors.append("Invalid e-mail address")
        if errors:
            return _fail(*errors)
        if owner is not None:
            self.owner = owner.strip()
        if email is not None:
            self.email = email
        return _ok(owner=self.owner, email=self.email)

    def get_balance(self):
        """View the balance (allowed in every state, including Frozen)."""
        return self.balance

    # ------------------------------------------------------------ history / CSV
    def get_history(self, start=None, end=None):
        """Return the transactions between ``start`` and ``end`` (both inclusive, both optional).

        Raises ``ValueError`` for unparsable dates or when ``start`` is after ``end``.
        """
        first = parse_date(start) if start is not None else None
        last = parse_date(end) if end is not None else None
        if (start is not None and first is None) or (end is not None and last is None):
            raise ValueError("Invalid date")
        if first and last and first > last:
            raise ValueError("Start date must not be after end date")
        return [
            tx
            for tx in self.transactions
            if (first is None or tx["date"] >= first) and (last is None or tx["date"] <= last)
        ]

    def export_csv(self, start=None, end=None):
        """Export the (optionally filtered) history as CSV text."""
        buffer = io.StringIO()
        writer = csv.writer(buffer, lineterminator="\n")
        writer.writerow(CSV_HEADER)
        for tx in self.get_history(start, end):
            writer.writerow(
                [
                    tx["date"].isoformat(),
                    tx["type"],
                    f"{tx['amount']:.2f}",
                    f"{tx['balance_after']:.2f}",
                    tx["description"],
                ]
            )
        return buffer.getvalue()
